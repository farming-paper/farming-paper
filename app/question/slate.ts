// This example is for an Editor with `ReactEditor` and `HistoryEditor`
import type { Token, Tokens } from "marked";
import { lexer } from "marked";
import type { Descendant } from "slate";
import type { CustomElement, CustomText } from "./types";

function convertMarkdownToText(token: Token): CustomText {
  switch (token.type) {
    case "text": {
      return { text: token.text };
    }

    case "strong": {
      const s = token as Tokens.Strong;
      const children = s.tokens.map(convertMarkdownToText);
      return { text: children.map((child) => child.text).join(""), bold: true };
    }

    default: {
      const s = token as Tokens.Generic;
      const children = s.tokens?.map(convertMarkdownToText);
      return { text: children?.map((child) => child.text).join("") ?? "" };
    }
  }
}

function convertMarkdownToElement(
  token: Token
): CustomElement | CustomElement[] | null {
  switch (token.type) {
    case "space": {
      return null;
    }

    case "paragraph": {
      const p = token as Tokens.Paragraph;
      const children = p.tokens.map(convertMarkdownToText);
      return { type: "paragraph", children };
    }

    case "list": {
      const list = token as Tokens.List;
      const result = list.items.map((item): CustomElement => {
        const children = item.tokens.map(convertMarkdownToText);
        const flatten = children
          .flat()
          .filter((child): child is CustomText => child !== null);
        return { type: "paragraph", children: flatten };
      });

      return result;
    }
  }

  return null;
}

export function convertMarkdownToDescendants(markdown: string): Descendant[] {
  const tokensList = lexer(markdown);
  return tokensList
    .map((token) => convertMarkdownToElement(token))
    .flat()
    .filter((child): child is CustomElement => child !== null);
}
