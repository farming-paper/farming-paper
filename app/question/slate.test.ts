import { describe, expect, test } from "vitest";
import { convertMarkdownToDescendants } from "./slate";

/** @see https://jaspervdj.be/lorem-markdownum/ */

describe("convertMarkdownToDescendants", () => {
  test("snapshot 1", () => {
    const doc = `Lorem markdownum errare mundi, sollicitumque numinis refers adfectu; cecidit nudumque ultimus: lucis invenit sanguine tuque; sors. Cum agitata [nymphae](#partu-sol-super), quem, gravem esse cum, amens Aeacides, **data** alto si, Ampycus erraverit. Hanc promptum captato. Tibi potentia nepotum per nec: haerentem vertar fatalis flammae queritur ex. Sacra secum ille superest voluntas unius autumnos, tamen, digna litore nec quid digerit naides, portant.

1. Vitisque exul quidem tamen Ammon leni canities
2. Ex partim iamdudum
3. Pando exitio nec credite adiit

Freta cum plumis mole mater serta rumpit Astyanax obvertit viscera suis. Sum abiit aetherium tribus, et adeunt solebat; quos aqua inamabile sileam. Consequiturque non ludit ego notavi Antigonen mater quid [illa facit](#coniunx).

Atque preme habes Antiphataeque aquae solverat, fuerit [anxia](#palustres-nec) capillos quisque! Amasse evocat, si undis erubuit parvos Io Iove. Nec crura puer agrosque, fortuna terrore, Leucippusque cursu in lustratum Thyesteis adversos.

Tunc est arida Achilles equarum hanc populo vocavit inpune auribus auxiliare pervia meritasque **et** adhuc; fata verbenis litore mors. Illis caelestia est, regno. Inpleratque furori si temeraria futuri, in viscere fera omni, adhaesit. Aliquem super, qua Idan quid facit, in illo ecce!

Suspectus multis mea gracili venerantur thalamos Iovisque intellege fronde vidit. Ut numina funeribus pontus, cortice vidisset, viridi; tum.

- Cum servitii venistis
- Maturae perdideris superant
- Hos lacrimas Trinacriam eurus Phene

Ore templum manu relatus, in [proelia](#vidi-premens-edocuit) stabant portitor montes consolenturne nova Vertumnumque ante per exstinctum! Equorum **quantum** aderat fixit ubi, de illis extulit fontis. Inposito igne venire noctis, vocant credensque operire!

Perde serpentis Oliaros transit Lucifer aquilam conplexibus vocatur; Myconon modo. Illis domus vacuae quae aliae lecto, unus cervis simul caesa fugere patulosque, dumque, vis Numam laetatur auget. Amphitrite sacrilegos **invecta ante liventia** ignibus, [sub vacent valido](#coget) humum capillos si fateri est quo mella aura: sunt. Laqueosque superatus, non Aricinae meae servant. Levant iuvenes, [Apolline ferrum](#terea-venatrixque-in); intonsum tardantis fervebat portis, paupertatemque rogant te sensimus.    
`;

    expect(convertMarkdownToDescendants(doc)).toMatchSnapshot();
  });

  test("snapshot 2", () => {
    const doc = `Lorem markdownum *montes* in patriis: casusque ornat sumit superi euntem dextera. Peremptis saepe!

1. Eripuit inpulit sit domito mihi saxis trepidumque
2. Deteriora serior cupit cesserat
3. Sedes causa Alcinoi habebatur linguae vires anima
4. Perenni deos
5. Quem visceraque decutit placet quae vulnus Cypriae

Duram sibi orbem, ut mersa edideras. Coeptaeque iacet, ardua tale iussis, quidve quae quaeque ligamina foedera.

Ire iuvenum optima. Non omnes ducere: visa nec imis **tot artes** subigebat pompas, amor? Placet quadrupes ossa attulit serpens proxima dixerat mihi vultu veteris delabere *me ferre* omnia *ille dicet*, Inachides.

1. Vero orbam aperta virum matrona tenet videtur
2. Quanto iuvenilior cadunt abeunt credita
3. Ferinae in reverentia Orion et cum
4. Sit Calydonia exercet umbris Procris mihi viscera
5. Formas nam dolendi ante spernitque

Adire parvumque ipsos, genitor dedit **tenebras** mea; properandum luctuque vel Pyrois *nec* tardius iussit, verbaque o. **Moverat** inter reverentia cursuque lacer, ius in agitata tamen subito, resoluta pendebat et vices refers: primisque. Nec Theseus veri hic considere nostri ambo gratia ultro, hos spisso mandata velamina ut maris licet. Quaeque est inventos i subit veniunt, at Quae *carcere nominis*. Audierit flammam testarique sorores *recondidit tempora* paries.

1. Ne et neglectos duabus sterilem color permansit
2. Arripit cui oculos ignibus exit his quamvis
3. Devia non aliisque
4. Fidensque simulacra sociis
5. Ad cum auras coercuit

Attollit niveis, mea conparentis, in ne faciem deposuit Stygii similis palus: est omnis ensis mugire accessit coepere. Adicit nullumque, undas nare vacuus in ne fila ad tua plura.`;

    expect(convertMarkdownToDescendants(doc)).toMatchSnapshot();
  });

  test("snapshot 3", () => {
    const doc = `Lorem markdownum. Aurum ante esse fugit gramina. Et quo crinem seque moenibus coniugis tu crede ipse esse tremendos reddat. Captivarum **deusque vestigia**, ora cepit ut arma gaudet valuere viscera cernunt, sua ego fugit velocius senex.

1. Est nec nymphas inpono inposito iuppiter verecundo
2. Inpubesque me regi Tyrrhena subit et expulit
3. Nepotum si praemia ante iunctura
4. Mox ut misit violenta

Quanto quaque ardore. Erigitur eiectat ripam innocuum, *erat locus* mandataque flamma caput te ille nurus? Optima et, quod atria ut perfida tua nulla grator. Pertulit vidi ferro arboreis et pater non [heros](#novus) estque, conligit a *nescio tangentia*.

- Meministis luminis sed
- Tantique gemitusque contigero moveret manifesta hac
- Profeci esse quid quem turaque monte
- Monte est ducit tum cernit iuventus ad
- Procubuit poscat

Solo fures sumitque inquam natis flamine spectat facinus. Si apud eras saxo gemit [toris](#suis) fuerunt ore vestra tantum tonitrua tot sustinuit ferunt quodque amore. Foramine durum, sic et minatur resolvo; mea veris ad radice omnia, ora terra continet et. Fontis color, tunc pro flammas en arbiter altis subit, manu loca firmatque parentum. Iram leves, et Icare: movimus caput et [fuisset](#hunc).

1. Anubis siste susurro
2. Naufragus factum condi vel
3. Patriam in moenia adiere
4. Mora turba fuisset possis erat usque profatur
5. Venit nobis erat vulnus advertere custos rogando
6. Et ossa ab spissatus artes refugitque an

Partu ratione pater: erat inter fecundam superi legati extrema. Iacent in Atque in, Pelagonaque ambit quo aura poenaque.

1. Nobis audit indignatus remove
2. Drya fallaces
3. Mersit sive hunc qua Memnonides Amnis cecidit
4. Loquentem dum
5. Legum et fuge iam
6. Rebar ad facit discidii

Videat es *Corinthiaci trepidantis* ferendo carpit relinquit imagine tuae sua colla o **nullius egit saxa** velatus **Peleus**. Lasso **et abesse septemplicis** lege matrem sub mihi vulgares, frena nec adfuit. Dumque ceciderunt radice. [Flavum Pleionesque](#dare-partibus) vocem naides quo Hymenaeus Euippe in lingua pictas; *nec*? Solita unde magna antra quique vitale rogat usus Bacchus [fatidicamque nocte](#ergo-exuvias).`;

    expect(convertMarkdownToDescendants(doc)).toMatchSnapshot();
  });
});
