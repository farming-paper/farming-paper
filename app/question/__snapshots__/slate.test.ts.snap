// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertMarkdownToDescendants > snapshot 1 1`] = `
[
  {
    "children": [
      {
        "text": "Lorem markdownum errare mundi, sollicitumque numinis refers adfectu; cecidit nudumque ultimus: lucis invenit sanguine tuque; sors. Cum agitata ",
      },
      {
        "text": "nymphae",
      },
      {
        "text": ", quem, gravem esse cum, amens Aeacides, ",
      },
      {
        "bold": true,
        "text": "data",
      },
      {
        "text": " alto si, Ampycus erraverit. Hanc promptum captato. Tibi potentia nepotum per nec: haerentem vertar fatalis flammae queritur ex. Sacra secum ille superest voluntas unius autumnos, tamen, digna litore nec quid digerit naides, portant.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Vitisque exul quidem tamen Ammon leni canities",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ex partim iamdudum",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Pando exitio nec credite adiit",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Freta cum plumis mole mater serta rumpit Astyanax obvertit viscera suis. Sum abiit aetherium tribus, et adeunt solebat; quos aqua inamabile sileam. Consequiturque non ludit ego notavi Antigonen mater quid ",
      },
      {
        "text": "illa facit",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Atque preme habes Antiphataeque aquae solverat, fuerit ",
      },
      {
        "text": "anxia",
      },
      {
        "text": " capillos quisque! Amasse evocat, si undis erubuit parvos Io Iove. Nec crura puer agrosque, fortuna terrore, Leucippusque cursu in lustratum Thyesteis adversos.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Tunc est arida Achilles equarum hanc populo vocavit inpune auribus auxiliare pervia meritasque ",
      },
      {
        "bold": true,
        "text": "et",
      },
      {
        "text": " adhuc; fata verbenis litore mors. Illis caelestia est, regno. Inpleratque furori si temeraria futuri, in viscere fera omni, adhaesit. Aliquem super, qua Idan quid facit, in illo ecce!",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Suspectus multis mea gracili venerantur thalamos Iovisque intellege fronde vidit. Ut numina funeribus pontus, cortice vidisset, viridi; tum.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Cum servitii venistis",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Maturae perdideris superant",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Hos lacrimas Trinacriam eurus Phene",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ore templum manu relatus, in ",
      },
      {
        "text": "proelia",
      },
      {
        "text": " stabant portitor montes consolenturne nova Vertumnumque ante per exstinctum! Equorum ",
      },
      {
        "bold": true,
        "text": "quantum",
      },
      {
        "text": " aderat fixit ubi, de illis extulit fontis. Inposito igne venire noctis, vocant credensque operire!",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Perde serpentis Oliaros transit Lucifer aquilam conplexibus vocatur; Myconon modo. Illis domus vacuae quae aliae lecto, unus cervis simul caesa fugere patulosque, dumque, vis Numam laetatur auget. Amphitrite sacrilegos ",
      },
      {
        "bold": true,
        "text": "invecta ante liventia",
      },
      {
        "text": " ignibus, ",
      },
      {
        "text": "sub vacent valido",
      },
      {
        "text": " humum capillos si fateri est quo mella aura: sunt. Laqueosque superatus, non Aricinae meae servant. Levant iuvenes, ",
      },
      {
        "text": "Apolline ferrum",
      },
      {
        "text": "; intonsum tardantis fervebat portis, paupertatemque rogant te sensimus.    ",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`convertMarkdownToDescendants > snapshot 2 1`] = `
[
  {
    "children": [
      {
        "text": "Lorem markdownum ",
      },
      {
        "text": "montes",
      },
      {
        "text": " in patriis: casusque ornat sumit superi euntem dextera. Peremptis saepe!",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Eripuit inpulit sit domito mihi saxis trepidumque",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Deteriora serior cupit cesserat",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Sedes causa Alcinoi habebatur linguae vires anima",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Perenni deos",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Quem visceraque decutit placet quae vulnus Cypriae",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Duram sibi orbem, ut mersa edideras. Coeptaeque iacet, ardua tale iussis, quidve quae quaeque ligamina foedera.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ire iuvenum optima. Non omnes ducere: visa nec imis ",
      },
      {
        "bold": true,
        "text": "tot artes",
      },
      {
        "text": " subigebat pompas, amor? Placet quadrupes ossa attulit serpens proxima dixerat mihi vultu veteris delabere ",
      },
      {
        "text": "me ferre",
      },
      {
        "text": " omnia ",
      },
      {
        "text": "ille dicet",
      },
      {
        "text": ", Inachides.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Vero orbam aperta virum matrona tenet videtur",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Quanto iuvenilior cadunt abeunt credita",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ferinae in reverentia Orion et cum",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Sit Calydonia exercet umbris Procris mihi viscera",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Formas nam dolendi ante spernitque",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Adire parvumque ipsos, genitor dedit ",
      },
      {
        "bold": true,
        "text": "tenebras",
      },
      {
        "text": " mea; properandum luctuque vel Pyrois ",
      },
      {
        "text": "nec",
      },
      {
        "text": " tardius iussit, verbaque o. ",
      },
      {
        "bold": true,
        "text": "Moverat",
      },
      {
        "text": " inter reverentia cursuque lacer, ius in agitata tamen subito, resoluta pendebat et vices refers: primisque. Nec Theseus veri hic considere nostri ambo gratia ultro, hos spisso mandata velamina ut maris licet. Quaeque est inventos i subit veniunt, at Quae ",
      },
      {
        "text": "carcere nominis",
      },
      {
        "text": ". Audierit flammam testarique sorores ",
      },
      {
        "text": "recondidit tempora",
      },
      {
        "text": " paries.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ne et neglectos duabus sterilem color permansit",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Arripit cui oculos ignibus exit his quamvis",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Devia non aliisque",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Fidensque simulacra sociis",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Ad cum auras coercuit",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Attollit niveis, mea conparentis, in ne faciem deposuit Stygii similis palus: est omnis ensis mugire accessit coepere. Adicit nullumque, undas nare vacuus in ne fila ad tua plura.",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`convertMarkdownToDescendants > snapshot 3 1`] = `
[
  {
    "children": [
      {
        "text": "Lorem markdownum. Aurum ante esse fugit gramina. Et quo crinem seque moenibus coniugis tu crede ipse esse tremendos reddat. Captivarum ",
      },
      {
        "bold": true,
        "text": "deusque vestigia",
      },
      {
        "text": ", ora cepit ut arma gaudet valuere viscera cernunt, sua ego fugit velocius senex.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Est nec nymphas inpono inposito iuppiter verecundo",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Inpubesque me regi Tyrrhena subit et expulit",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Nepotum si praemia ante iunctura",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Mox ut misit violenta",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Quanto quaque ardore. Erigitur eiectat ripam innocuum, ",
      },
      {
        "text": "erat locus",
      },
      {
        "text": " mandataque flamma caput te ille nurus? Optima et, quod atria ut perfida tua nulla grator. Pertulit vidi ferro arboreis et pater non ",
      },
      {
        "text": "heros",
      },
      {
        "text": " estque, conligit a ",
      },
      {
        "text": "nescio tangentia",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Meministis luminis sed",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Tantique gemitusque contigero moveret manifesta hac",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Profeci esse quid quem turaque monte",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Monte est ducit tum cernit iuventus ad",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Procubuit poscat",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Solo fures sumitque inquam natis flamine spectat facinus. Si apud eras saxo gemit ",
      },
      {
        "text": "toris",
      },
      {
        "text": " fuerunt ore vestra tantum tonitrua tot sustinuit ferunt quodque amore. Foramine durum, sic et minatur resolvo; mea veris ad radice omnia, ora terra continet et. Fontis color, tunc pro flammas en arbiter altis subit, manu loca firmatque parentum. Iram leves, et Icare: movimus caput et ",
      },
      {
        "text": "fuisset",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Anubis siste susurro",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Naufragus factum condi vel",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Patriam in moenia adiere",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Mora turba fuisset possis erat usque profatur",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Venit nobis erat vulnus advertere custos rogando",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Et ossa ab spissatus artes refugitque an",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Partu ratione pater: erat inter fecundam superi legati extrema. Iacent in Atque in, Pelagonaque ambit quo aura poenaque.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Nobis audit indignatus remove",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Drya fallaces",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Mersit sive hunc qua Memnonides Amnis cecidit",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Loquentem dum",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Legum et fuge iam",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Rebar ad facit discidii",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Videat es ",
      },
      {
        "text": "Corinthiaci trepidantis",
      },
      {
        "text": " ferendo carpit relinquit imagine tuae sua colla o ",
      },
      {
        "bold": true,
        "text": "nullius egit saxa",
      },
      {
        "text": " velatus ",
      },
      {
        "bold": true,
        "text": "Peleus",
      },
      {
        "text": ". Lasso ",
      },
      {
        "bold": true,
        "text": "et abesse septemplicis",
      },
      {
        "text": " lege matrem sub mihi vulgares, frena nec adfuit. Dumque ceciderunt radice. ",
      },
      {
        "text": "Flavum Pleionesque",
      },
      {
        "text": " vocem naides quo Hymenaeus Euippe in lingua pictas; ",
      },
      {
        "text": "nec",
      },
      {
        "text": "? Solita unde magna antra quique vitale rogat usus Bacchus ",
      },
      {
        "text": "fatidicamque nocte",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
]
`;
